File: solutions/java/run-length-encoding/1/src/main/java/RunLengthEncoding.java
import java.util.regex.*;

class RunLengthEncoding {

    String encode(String data) {
        Pattern pattern=Pattern.compile("(.)\\1+"); 
            //Note: NOT Pattern.CASE_INSENSITIVE
        Matcher matcher=pattern.matcher(data);
        return matcher.replaceAll(matchResult ->  matchResult.group().length() + matchResult.group(1));
    }

    String decode(String data) {
         Pattern pattern=Pattern.compile("(\\d+)(.)");
             // Note: NOT Pattern.CASE_INSENSITIVE
        Matcher matcher=pattern.matcher(data);
        return matcher.replaceAll(matchResult ->  matchResult.group(2).repeat(Integer.parseInt(matchResult.group(1))));
    }

}

File: solutions/kotlin/run-length-encoding/1/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String =
        input.replace(Regex("(.)\\1+")){
            it.value.length.toString() + it.groupValues[1]}
        
    fun decode(input: String): String =
        input.replace(Regex("(\\d+)(.)")){
            it.groupValues[2]
            .repeat(Integer.parseInt(it.groupValues[1]))
        }
}
